src/ClientHandle.cpp

```diff
@@ -1243,13 +1246,32 @@ void cClientHandle::HandleBlockDigFinished(int a_BlockX, int a_BlockY, int a_Blo
 		}
 	}
 
+	if (!m_Player->IsGameModeCreative() && !cBlockInfo::IsOneHitDig(a_OldBlock))
+	{
+		// Fix for very fast tools.
+		m_BreakProgress += m_Player->GetPlayerRelativeBlockHardness(a_OldBlock);
+		if (m_BreakProgress < FASTBREAK_PERCENTAGE)
+		{
+			LOGD("Break progress of player %s was less than expected: %f < %f\n", m_Player->GetName().c_str(), m_BreakProgress * 100, FASTBREAK_PERCENTAGE * 100);
+			// AntiFastBreak doesn't agree with the breaking. Bail out. Send the block back to the client, so that it knows:
+			m_Player->GetWorld()->SendBlockTo(a_BlockX, a_BlockY - 1, a_BlockZ, m_Player);  // Bug with doors and things like that.
+			m_Player->GetWorld()->SendBlockTo(a_BlockX, a_BlockY, a_BlockZ, m_Player);
+			m_Player->GetWorld()->SendBlockTo(a_BlockX, a_BlockY + 1, a_BlockZ, m_Player);  // Bug with doors and things like that.
```

Comment:
> That's also an issue if plugins want to cancel block breaking so we should create a function in cPlayer or somewhere which resends the required things around a broken block. 

It looks like `cClientHandle::HandleBlockDigFinished` does this; it currently resends the block that was broken and the player's position if it was canceled (look for `CallHookPlayerBreakingBlock`).

Craftbukkit does a similar thing, except it always sends a set block to air packet and then will send an additional set block packet resseting it _only_ if the event is canceled.  It also resends the block's tile entity packet if it has one.  No handling of doors that I can see.  I'll check the bug tracker to see if there's stuff about doors being poorly handled in a bit.

> That's also an issue if plugins want to cancel block breaking so we should create a function in cPlayer or somewhere which resends the required things around a broken block.



> Do you know what the vanilla server does if the block Is broken too fast?

I mentioned this on #3409.  As observable through [MC-5694](https://bugs.mojang.com/browse/MC-5694), the block isn't always resent.  (It might be resent in some cases, and I actually think it is due to "popping back" behavior I've seen, but it isn't always).  In any case the vanilla server probably isn't a good model for this behavior.